neural_chat/README.md

```diff
@@ -1,214 +1,157 @@
+## Advance Topics
 
+### Plugins
 
-## Finetuning
+NeuralChat introduces the `plugin` machenism which integrate a lot of useful LLM utils and features to augment the chatbot's capability. Such plugins are applied in the chatbot pipeline for inference.
 
-Finetune the pretrained large language model (LLM) with the instruction-following dataset for creating the customized chatbot is very easy for NeuralChat.
+Belows are supported plugins.
 
-**command line experience**
+- [knowledge retrieval](./pipeline/plugins/retrievers/)
 
-```shell
-neuralchat finetune --base_model "meta-llama/Llama-2-7b-chat-hf" --config pipeline/finetuning/config/finetuning.yaml
-```
+    Knowledge retrieval consists of document indexing for efficient retrieval of relevant information, including Dense Indexing based on LangChain and Sparse Indexing based on fastRAG, document rankers to prioritize the most relevant responses.
 
+- [query caching](./pipeline/plugins/caching/)
 
-**Python API experience**
+    Query caching enables the fast path to get the response without LLM inference and therefore improves the chat response time
 
-```python
->>> from neural_chat import FinetuningConfig
->>> from neural_chat import finetune_model
->>> finetune_cfg = FinetuningConfig()
->>> finetuned_model = finetune_model(finetune_cfg)
-```
+- [prompt optimization](./pipeline/plugins/prompts/)
 
-## Quantization
+    Prompt optimization supports auto prompt engineering to improve user prompts.
 
-NeuralChat provides three quantization approaches respectively (PostTrainingDynamic, PostTrainingStatic, QuantAwareTraining) based on [Intel® Neural Compressor](https://github.com/intel/neural-compressor).
+- [memory controller](./pipeline/plugins/memory/)
 
-**command line experience**
+    Memory controller enables the efficient memory utilization.
 
-```shell
-neuralchat optimize --base_model "meta-llama/Llama-2-7b-chat-hf" --config pipeline/optimization/config/optimization.yaml
-```
+- [safety checker](./pipeline/plugins/security/)
 
+    Safety checker enables the sensitive content check on inputs and outputs of the chatbot.
 
-**Python API experience**
+User could enable, disable, and even change the default behavior of all supported plugins like below
 
 ```python
->>> from neural_chat import OptimizationConfig
->>> from neural_chat import optimize_model
->>> opt_cfg = OptimizationConfig()
->>> optimized_model = optimize_model(opt_cfg)
-```
-
-
-<a name="quickstartserver"></a>
-# Quick Start Server
-
-Users can have a try of NeuralChat server with [NeuralChat Server Command Line](./server/README.md).
+from neural_chat import build_chatbot, PipelineConf, plugins
 
+plugins.retrival.enabe = True
+plugins.retrival.path = '/path/to/user/doc'
+conf = PipelineConf(plugins=plugins)
```

Comment:
PipelineConf(plugins=plugins) ====> PipelineConf(plugin=plugins)

Reply:
If plugin = plugins, then we should use plugins = plugins